src/serial.c

```diff
@@ -0,0 +1,230 @@
+
+            if (fifo_is_empty(&priv->rx_buf)) {
+                priv->lsr &= ~UART_LSR_DR;
+                serial_update_irq(s);
+            }
+        }
+        break;
+    case UART_IER:
+        if (priv->lcr & UART_LCR_DLAB)
+            IO_WRITE8(data, priv->dlm);
+        else
+            IO_WRITE8(data, priv->ier);
+        break;
+    case UART_IIR:
+        IO_WRITE8(data, priv->iir | 0xc0);  // 0xc0 stands for FIFO enabled
+        break;
+    case UART_LCR:
+        IO_WRITE8(data, priv->lcr);
+        break;
+    case UART_MCR:
+        IO_WRITE8(data, priv->mcr);
+        break;
+    case UART_LSR:
+        IO_WRITE8(data, priv->lsr);
+        break;
+    case UART_MSR:
+        IO_WRITE8(data, priv->msr);
+        break;
+    case UART_SCR:
+        IO_WRITE8(data, priv->scr);
+        break;
+    default:
+        break;
+    }
+    pthread_mutex_unlock(&s->lock);
+}
+
+static void serial_out(serial_dev_t *s, uint16_t offset, void *data)
+{
+    struct serial_dev_priv *priv = (struct serial_dev_priv *) s->priv;
+
+    pthread_mutex_lock(&s->lock);
+
+    switch (offset) {
+    case UART_TX:
+        if (priv->lcr & UART_LCR_DLAB) {
+            priv->dll = IO_READ8(data);
+        } else {
+            priv->lsr |= (UART_LSR_TEMT | UART_LSR_THRE);  // flush TX
+            putchar(((char *) data)[0]);
+            fflush(stdout);
+            serial_update_irq(s);
+        }
+        break;
+    case UART_IER:
+        if (!(priv->lcr & UART_LCR_DLAB)) {
+            priv->ier = IO_READ8(data);
+            serial_update_irq(s);
+        } else {
+            priv->dlm = IO_READ8(data);
+        }
+        break;
+    case UART_FCR:
+        priv->fcr = IO_READ8(data);
+        break;
+    case UART_LCR:
+        priv->lcr = IO_READ8(data);
+        break;
+    case UART_MCR:
+        priv->mcr = IO_READ8(data);
+        break;
+    case UART_LSR:  // factory test
+    case UART_MSR:  // not used
+        break;
+    case UART_SCR:
+        priv->scr = IO_READ8(data);
+        break;
+    default:
+        break;
+    }
+    pthread_mutex_unlock(&s->lock);
+}
+
+void serial_init(serial_dev_t *s)
+{
+    *s = (serial_dev_t){
+        .priv = (void *) &serial_dev_priv,
+    };
+
+    pthread_mutex_init(&s->lock, NULL);
+    s->infd = STDIN_FILENO;
+    // create a thread which accepts serial input
+    pthread_t tid;
+    pthread_create(&tid, NULL, (void *) serial_console, (void *) s);
+}
+
+void serial_handle(serial_dev_t *s,
+                   int direction,
+                   uint16_t port,
+                   uint8_t size,
+                   uint32_t count,
+                   void *data)
+{
+    bool is_write = (direction == KVM_EXIT_IO_OUT);
+    uint16_t offset = port - COM1_PORT_BASE;
+
+    while (count--) {
+        if (is_write)
+            serial_out(s, offset, data);
+        else
+            serial_in(s, offset, data);
+
+        data += size;
+    }
+}
+
+void serial_exit(serial_dev_t *s)
```

Comment:
Would you release the resources here?

Reply:
I think we should reclaim mutex here. The resource of the thread (from `pthread_create`) should be released by the way?

Reply:
The resource mutex should be released here. I think I should also reclaim the resource of thread by the way?

Reply:
I was worried about the thread created in function `serial_init`. Maybe the invocation of `pthread_cancel` can help from the perspective of resource management.

Reply:
I'm not sure if `pthread_cancel` is a good solution because we may need a cancellation point in our thread (?). How about we just maintain a global state and loop according to it, instead of using infinite loop for our console thread. Then, if `serial_exit` is executed, the main thread (or maybe the signal handler in the future because we usually use `Ctrl+C` to stop our program) can change the state and let the console thread exit itself, so we can `pthread_join` it.

Reply:
We can atomic operation to maintain the global state, so I think this could be an easier solution?

Reply:
Affirmative. Let's try global state tracking with atomics. Also, `atexit` and signal handler may be used for such scenario.